lib/semantics/tools.h

```diff
@@ -175,6 +175,15 @@ const DeclTypeSpec &FindOrInstantiateDerivedType(Scope &, DerivedTypeSpec &&,
     SemanticsContext &, DeclTypeSpec::Category = DeclTypeSpec::TypeDerived);
 void ProcessParameterExpressions(DerivedTypeSpec &, evaluate::FoldingContext &);
 
+// Check to see if a variable being redefined is a DO variable.  If so, emit
+// a message
+void SayDoVarRedefine(
+    const Symbol &, SemanticsContext &, const parser::CharBlock &);
+void CheckDoVarRedefine(
+    const Symbol &, SemanticsContext &, const parser::CharBlock &);
+void CheckDoVarRedefine(const parser::Variable &, SemanticsContext &);
+void CheckDoVarRedefine(const parser::Name &, SemanticsContext &);
```

Comment:
It might make more sense to make these be member functions of `SemanticContext` as they depend on the active do variable there. It seems like that would make the new API simpler: just Activate/DeactivateDoVariable and CheckDoVarRedefine.

Reply:
Good suggestion.  I'll do that.